lib/galaxy/managers/datasets.py

```diff
@@ -684,6 +685,20 @@ def deserialize_metadatum(self, dataset_assoc, key, val, **context):
         # ...?
         return unwrapped_val
 
+    def deserialize_datatype(self, item, key, val, **context):
+        if not item.datatype.allow_datatype_change:
+            raise exceptions.RequestParameterInvalidException("The current datatype does not allow datatype changes.")
+        if not self.app.datatypes_registry.get_datatype_by_extension(val):
+            raise exceptions.RequestParameterInvalidException("The target datatype does not exist.")
+        if not self.app.datatypes_registry.get_datatype_by_extension(val).allow_datatype_change:
+            raise exceptions.RequestParameterInvalidException("The target datatype does not allow datatype changes.")
+        if not DatasetAssociationManager(self.app).ok_to_edit_metadata(item.dataset_id):
+            raise exceptions.RequestParameterInvalidException("Dataset metadata could not be updated, probably because it is not in an 'ok' state.")
```

Comment:
That's not an invalid parameter in the request, so I would raise a normal exception (I don't think we have any specific Galaxy exception), which should also return a 500 HTTP error code (this is in fact more a server than user error).

```suggestion
            raise Exception("Dataset metadata could not be updated, probably because it is not in an 'ok' state.")
```